### ISSUE TITLE
Need help to understand the code

### ISSUE THREAD

**Issue by iamxpy** [zysszy/TreeGen#5]
First, I would like to thank you for the publication of the source code, which makes it much easier for me to understand the details of the model. I have read several key files of the project, including `resolve_data.py`, `code_generate_model.py` and `run.py`, and now I am kind of stuck on the file `predict.py`. I have seversal questions and need some help.

(1) **What does the symbol `^` mean**? It appears both in the input file of training procedure (`train_trans.txt`) and the output file of inference (e.g. 'ATIS/out/0.txt'), and the code in `predict.py` insert or remove the symbol many times. 

(2) **How to understand the content of the output files of `predict.py`(e.g. `ATIS/out/0.txt`)?** According to the method `BeamSearch()` in `predict.py`, I think every 2 lines is a beam, with the first line showing the generated tree and the second line is the probability of the tree. But I failed to understand how the trees are organized into these sequences (with so many`^`), and what made me more confused, the "probability" are all negative numbers. I have attached a picture showing the content of `0.txt`.

![image](https://user-images.githubusercontent.com/20579310/96689921-9d0fd180-13b5-11eb-81b6-a46ac990d356.png)

(3)**Which part of this repository is not complete?** According to README, "This repository is not a complete one".

(4) **The number of training epochs is set to `pretrain_times = 100000`, is it necessary to use such a big value to achieve the results in the paper?** After running for a while, I found out that after 100 epochs, the accuracy is 0.98+, and 0.99+ after 500 epochs, so I just stopped training.

Thank you for your help in advance!

**Comment by iamxpy** [comment]
After debugging for several test samples, I think I understand the symbol `^` in `ATIS/out/n.txt` now. According `J_AddSon()` method in `predict.py`, the symbol `^` is used to match the corresponding key word, we can add some brackets to make it more clear. 

For example, the sequence 

```root Lambda variable $0 ^ ^ type e ^ ^ body Apply predicate airport ^ ^ arguments Variable variable $0 ^ ^ ^ End ^ ^ ^ ^ ^ ^``` 

equals to 

```
(root (Lambda (variable ($0 ^) ^) (type (e ^) ^) (body (Apply (predicate (airport ^) ^) (arguments (Variable (variable ($0 ^) ^) ^) (End ^) ^) ^) ^) ^) ^)
```
, and it equals to the following tree:

![image](https://user-images.githubusercontent.com/20579310/96729309-98641100-13e7-11eb-9e5f-b735bd98e172.png)

**Comment by zysszy** [comment]
Thank you for your attention~

(1) What does the symbol ^ mean?  
Yeah, you got it!  Your understanding is absolutely right~

(2) How to understand the content of the output files of predict.py(e.g. ATIS/out/0.txt)?
```how the trees are organized into these sequences``` I think the answer is same to the question 1.
```the "probability" are all negative numbers``` The probability here is the ```log``` of its percentage. 

(3) Which part of this repository is not complete?
I am sorry for it. We have not provided the experiments of ```GEO``` and ```ATIS``` in the paper.

(4) The number of training epochs is set to pretrain_times = 100000, is it necessary to use such a big value to achieve the results in the paper?
It is not necessary to use such a large value. You can stop training when the accuracy is no longer growing. 

I hope my answers can help you understand our project. 
If you have any questions, please feel free to tell me~

Zeyu

**Comment by iamxpy** [comment]
> Thank you for your attention~
> 
> (1) **What does the symbol ^ mean?**
> Yeah, you got it! Your understanding is absolutely right~
> 
> (2) **How to understand the content of the output files of predict.py(e.g. ATIS/out/0.txt)?**
> `how the trees are organized into these sequences` I think the answer is same to the question 1.
> `the "probability" are all negative numbers` The probability here is the `log` of its percentage.
> 
> (3) **Which part of this repository is not complete?**
> I am sorry for it. We have not provided the experiments of `GEO` and `ATIS` in the paper.
> 
> (4) **The number of training epochs is set to pretrain_times = 100000, is it necessary to use such a big value to achieve the results in the paper?**
> It is not necessary to use such a large value. You can stop training when the accuracy is no longer growing.
> 
> I hope my answers can help you understand our project.
> If you have any questions, please feel free to tell me~
> 
> Zeyu

Thanks for your quick reply, I still have some questions.

(1) **Confusion about the meaning of "Javaoutput.Probability"**. 

Specifically, I get confused about the following expression in `predict.py`:

```python
JavaOutNext.Probility = (JavaOut.Probility * math.pow(len(JavaOut.RuleList), apa) + math.log(max(1e-10, res[i]))) / math.pow(len(JavaOut.RuleList) + 1, apa)
```

Suppose we decode rule `A` at time step 1, and get the corresponding probability `P(A)`(the output of softmax function), and decode rule `B` at time step 2 with `P(B)`, and so on.

I notice that, at time step 1, the expression above equals to `log(P(A))` because `len(JavaOut.RuleList)` is `0`. At time step 2, the expression equals to 

![image](https://user-images.githubusercontent.com/20579310/96976212-2528df00-154e-11eb-97ee-763a72771f7f.png)

in which `len` is `1`.

I hope my calculation is not wrong, and could you explain more about "the log of its percentage"?

(2) **Why do we need the files `Tree_Rule.in` and `Tree_Feature.out`**? Why can't we just simply pass the content of these files (as a list or dict) to the methods which need them? And most importantly I failed to understand the content of `Tree_Feature.out`, namly the following snippet:

```python
fw.write(out.replace(" node_gen ^", "") + "\n")
fw.write(node_par[0] + "\n")
fw.write(node_par[1] + "\n")
fw.write(node_par[2] + "\n")
fw.write(out.replace(" End ^", "") + "\n")
fw.write(out.replace(" End ^", "") + "\n")
fw.write("1\n")
fw.write("1\n")
fw.write(out.replace(" End ^", "") + "\n")
fw.write("1\n")
fw.write(str(father_index_now) + "\n")
```

Hoping for your reply!

**Comment by zysszy** [comment]
(1) **Confusion about the meaning of "Javaoutput.Probability".**
```apa``` denotes the ```alpha```, which is a length penalty in ```beam search```. Details are in https://d2l.ai/chapter_recurrent-modern/beam-search.html (9.8.4).

(2) **Why do we need the files Tree_Rule.in and Tree_Feature.out?**
Actually, we don't need these two files. This ```python``` code is converted from ```Java```, and I left these two files to avoid introducing new bugs. With the development of this research, we have left a lot of useless code, which is hard to remove, in this project.

Zeyu

**Comment by iamxpy** [comment]
@zysszy Many thanks for helping me out so much!

**Comment by iamxpy** [comment]
@zysszy Sorry to bother you again, I have another question. As you've mentioned in #3 

> we always expand the leftmost non-terminal node until all leaf nodes are terminals

I can confirm that this is true when doing inference in `predict.py`, but maybe I missed something and I think it is not the case in the training process. For example, the whole AST of the first training sample (in `train_trans.txt` and `train_tree.txt`) looks like this:


![微信图片_20201027204223](https://user-images.githubusercontent.com/20579310/97303485-68a88380-1895-11eb-8bd6-4d53d4b2de5a.jpg)

The numbers beside every non-terminal nodes indicate their order which decide which tokens they can pay attention to under the control of the `antimask`, a lower triangular matrix. In this example, after expanding 8th node `args` (short for `arguments`), the model expands 9th node `args` instead of the node `Var` (the leftmost non-terminal node after expanding the first `args`). 

What am I doing wrong here? Thanks!

**Comment by zysszy** [comment]
Sorry for the late reply.

The Node ```End```, which denotes the end of the child nodes of ```Apply```, is a child node of the node ```Apply```. Thus, the 8th node ```args``` is still the leftmost non-terminal node.

Zeyu

**Comment by iamxpy** [comment]
I still feel confused about **the Node `End` and the meaning of `leftmost`**.

First of all, I would like to clarify that the data of the first example is as follows.

```
airport
root Lambda variable $0 $0^ variable^ type e e^ type^ body Apply predicate airport airport^ predicate^ arguments Variable variable node_gen node_gen^ variable^ Variable^ arguments^ Apply^ body^ Lambda^ root^
Unknown root Lambda variable variable^ Lambda^ Lambda type type^ Lambda^ Lambda body Apply predicate predicate^ Apply^ Apply arguments Variable variable variable^ Variable^ arguments^ Apply^ body^ Lambda^ root^ Unknown^
Unknown Unknown root Lambda Lambda^ root^ root Lambda Lambda^ root^ root Lambda body Apply Apply^ body^ body Apply arguments Variable Variable^ arguments^ Apply^ body^ Lambda^ root^ Unknown^ Unknown^
variable  Variable  arguments  Apply  body  Lambda  root
1 2 3 4 5 6 7 8 9 10 3
0 1 2 3 4 5 6 7 8 9 2 
9
-1 0 1 1 1 4 5 5 5 7 9 
```

(1) Question about the Node `End`

> The Node `End`, which denotes the end of the child nodes of `Apply`, is a child node of the node `Apply`

There is no rule with `Apply` on the left side and `End` on the right side in file `ATIS/Rule.txt`, however the rule `arguments true End` exists and it is exactly the rule used in the first example (`9` in the 6th line indicates this).

**Comment by zysszy** [comment]
I found the figure is wrong. 8th node ```args``` and 9th `args` are the same node. Thus, we first expand the 8th node with the Rule `7` (`arguments true Variable`). Then, all children of the 8th node were generated, and we expand the 8th node with the Rule `8` (`arguments true End`). So, the next step is to expand the 10th node `Var`.

> @zysszy Sorry to bother you again, I have another question. As you've mentioned in #3
> 
> > we always expand the leftmost non-terminal node until all leaf nodes are terminals
> 
> I can confirm that this is true when doing inference in `predict.py`, but maybe I missed something and I think it is not the case in the training process. For example, the whole AST of the first training sample (in `train_trans.txt` and `train_tree.txt`) looks like this:
> 
> ![微信图片_20201027204223](https://user-images.githubusercontent.com/20579310/97303485-68a88380-1895-11eb-8bd6-4d53d4b2de5a.jpg)
> 
> The numbers beside every non-terminal nodes indicate **their order which decide which tokens they can pay attention to** under the control of the `antimask`, a lower triangular matrix. In this example, after expanding 8th node `args` (short for `arguments`), the model expands 9th node `args` instead of the node `Var` (the leftmost non-terminal node after expanding the first `args`).
> 
> What am I doing wrong here? Thanks!

**Comment by iamxpy** [comment]
> I found the figure is wrong. 8th node args and 9th args are the same node. Thus, we first expand the 8th node with the Rule 7 (arguments true Variable). Then, all children of the 8th node were generated, and we expand the 8th node with the Rule 8 (arguments true End). So, the next step is to expand the 10th node Var (the leftmost non-terminal node).

How to understand the last line of the first example, `-1 0 1 1 1 4 5 5 5 7 9`? Based on the code, I think after adding `1` to every number and appending a `-1` to the left, we can get the [parent array representation](https://www.geeksforgeeks.org/construct-a-binary-tree-from-parent-array-representation/) of the AST, i.e. `-1 0 1 2 2 2 5 6 6 6 8 10`. We can see that the 6th node has 3 children including the 8th node and the 9th node. I built the tree based on the parent array(the last line of every example) and the rules(the 6th line of every example), what am I doing wrong? Thx!

**Comment by zysszy** [comment]
> > I found the figure is wrong. 8th node args and 9th args are the same node. Thus, we first expand the 8th node with the Rule 7 (arguments true Variable). Then, all children of the 8th node were generated, and we expand the 8th node with the Rule 8 (arguments true End). So, the next step is to expand the 10th node Var (the leftmost non-terminal node).
> 
> How to understand the last line of the first example, `-1 0 1 1 1 4 5 5 5 7 9`? Based on the code, I think after adding `1` to every number and appending a `-1` to the left, we can get the [parent array representation](https://www.geeksforgeeks.org/construct-a-binary-tree-from-parent-array-representation/) of the AST, i.e. `-1 0 1 2 2 2 5 6 6 6 8 10`. We can see that the 6th node has 3 children including the 8th node and the 9th node. I built the tree based on the parent array(the last line of every example) and the rules(the 6th line of every example), what am I doing wrong? Thx!

Yeah, you are right if you remove the node `start` and use the node `root` as the 0th node~

**Comment by iamxpy** [comment]
> > > I found the figure is wrong. 8th node args and 9th args are the same node. Thus, we first expand the 8th node with the Rule 7 (arguments true Variable). Then, all children of the 8th node were generated, and we expand the 8th node with the Rule 8 (arguments true End). So, the next step is to expand the 10th node Var (the leftmost non-terminal node).
> > 
> > 
> > How to understand the last line of the first example, `-1 0 1 1 1 4 5 5 5 7 9`? Based on the code, I think after adding `1` to every number and appending a `-1` to the left, we can get the [parent array representation](https://www.geeksforgeeks.org/construct-a-binary-tree-from-parent-array-representation/) of the AST, i.e. `-1 0 1 2 2 2 5 6 6 6 8 10`. We can see that the 6th node has 3 children including the 8th node and the 9th node. I built the tree based on the parent array(the last line of every example) and the rules(the 6th line of every example), what am I doing wrong? Thx!
> 
> Yeah, you are right if you remove the node `start` and use the node `root` as the 0th node~

I understand that the 6th line of every sample corresponds to the tree without `start`. But let us just use the the tree with `start` so we don't have to change(minus 1) the indices. I still don't understand why can we conclude that `8th node args and 9th args are the same node`.

**Comment by zysszy** [comment]
> > > > I found the figure is wrong. 8th node args and 9th args are the same node. Thus, we first expand the 8th node with the Rule 7 (arguments true Variable). Then, all children of the 8th node were generated, and we expand the 8th node with the Rule 8 (arguments true End). So, the next step is to expand the 10th node Var (the leftmost non-terminal node).
> > > 
> > > 
> > > How to understand the last line of the first example, `-1 0 1 1 1 4 5 5 5 7 9`? Based on the code, I think after adding `1` to every number and appending a `-1` to the left, we can get the [parent array representation](https://www.geeksforgeeks.org/construct-a-binary-tree-from-parent-array-representation/) of the AST, i.e. `-1 0 1 2 2 2 5 6 6 6 8 10`. We can see that the 6th node has 3 children including the 8th node and the 9th node. I built the tree based on the parent array(the last line of every example) and the rules(the 6th line of every example), what am I doing wrong? Thx!
> > 
> > 
> > Yeah, you are right if you remove the node `start` and use the node `root` as the 0th node~
> 
> I understand that the 6th line of every sample corresponds to the tree without `start`. But let us just use the the tree with `start` so we don't have to change(minus 1) the indices. I still don't understand why can we conclude that `8th node args and 9th args are the same node`.

Hmmm... Sorry, it is a bug..... 
I think you are right. I set a wrong parent node, which is a grandparent node, to this rule.
This bug is not an important bug. It will not affect the effectiveness of TreeGen.

Thank you for this.

Zeyu

**Comment by iamxpy** [comment]
@zysszy Thank you soooo much for your valuable time!

**Comment by zhaodezhu111** [comment]
@zysszy I want to ask if this method of AST node traversal into a sequence is from that paper. Is there a paper that officially introduces this method? My current work uses this AST traversal method, and I want to quote the reference of this method.

**Comment by zysszy** [comment]
> @zysszy I want to ask if this method of AST node traversal into a sequence is from that paper. Is there a paper that officially introduces this method? My current work uses this AST traversal method, and I want to quote the reference of this method.

We directly use the traditional preorder traversal algorithm.  Maybe you can refer to the Wikipedia (https://en.wikipedia.org/wiki/Tree_traversal).

**Comment by zhaodezhu111** [comment]
OK, thank you for your reply！
